torch/quantization/fx/quantize.py

```diff
@@ -197,16 +197,27 @@ def is_observed(input_arg):
             assert matched_nodes is not None
             input_node = matched_nodes[-1]  # first node in the sequence
 
-            def input_is_observed(arg):
-                return (isinstance(arg, Node) and
-                        arg.name in observed_node_names_set)
-            # This is checking if one of the argument of add/mul
-            # is an observed node
-            # If both of the inputs are number,
-            # we will not consider the output to be observed
-            if (input_is_observed(input_node.args[0]) or
-                    input_is_observed(input_node.args[1])):
-                observed_node_names_set.add(node.name)
+            if not model.training:
+                # for post-training quantization, we only observe the output
+                # if one the inputs is observed
+                def input_is_observed(arg):
+                    return (isinstance(arg, Node) and
+                            arg.name in observed_node_names_set)
+                # This is checking if one of the argument of add/mul
+                # is an observed node
+                # If both of the inputs are number,
+                # we will not consider the output to be observed
+                if (input_is_observed(input_node.args[0]) or
+                        input_is_observed(input_node.args[1])):
+                    observed_node_names_set.add(node.name)
+            else:
+                # for QAT, we always insert a fake_quant after add/mul
+                new_observer = qconfig.activation()
+                insert_observer(
```

Comment:
I think we'll need to use the same fake quantize module as the input here to properly simulate quantization since output of quantized::add with scalar input inherits quantization parameters from input

Reply:
Not sure I follow: The current logic inserts observers as long as one of the inputs is a tensor. why is this a problem for QAT?